### PR TITLE
support instantiating money from float

### PR DESCRIPTION
**Pull request by bendavies**
This is mainly in response to https://github.com/moneyphp/money/issues/408

Any thoughts?

### CODE REVIEW

File: src/Number.php [hunk 1]
```diff
@@ -117,7 +117,7 @@ public static function fromString($number)
     /**
      * @param float $floatingPoint
      *
-     * @return Number
+     * @return self
```
Comment (sagikazarmark) [inline src/Number.php]: I tend to prefer using the class name in return type hints to indicate a new object is going to be returned. At least for consistency, this should be reverted.

File: src/Money.php [hunk 2]
```diff
@@ -49,20 +49,24 @@
...
+     * @param Currency         $currency
      *
      * @throws \InvalidArgumentException If amount is not integer
      */
     public function __construct($amount, Currency $currency)
     {
         if (filter_var($amount, FILTER_VALIDATE_INT) === false) {
-            $numberFromString = Number::fromString($amount);
-            if (!$numberFromString->isInteger()) {
+            if (is_float($amount)) {
```
Comment (sagikazarmark) [inline src/Money.php]: Not sure this will work as intended. What if `$amount == (float)PHP_INT_MAX+1` ?

File: src/Number.php (hunk 1 above)
Comment (bendavies) [reply src/Number.php]: i was making consistent with `fromString` above. i'll revert this and you can handle in another PR if you wish.

File: src/Money.php (hunk 2 above)
Comment (bendavies) [reply src/Money.php]: this works fine.

Comment (bendavies) [reply src/Money.php]: sorry, i misread. yes this doesn't work. what should be the behaviour/error here?

**Comment by bendavies** [comment]
@frederikbosch there are two tests, one from a float, and one from scientific notation. (and i was just testing the maximum float representable by php fyi).

btw, why is there `MoneySpec` and `MoneyTest`? There seems to be some overlap.

**Comment by frederikbosch** [comment]
Well I am simply against accepting one more type in the constructor. Use a named public static constructor it.

**Comment by bendavies** [comment]
@frederikbosch sorry, i missed what you were objecting to, hence my non nonsense reply.

**Comment by sagikazarmark** [comment]
I second @frederikbosch 's opinion. Would `Money::fromNumber(Number::fromFloat(3.5))` work for you @bendavies ?

**Comment by bendavies** [comment]
@sagikazarmark hey! that doesn't add anything to me over my current solution which is:`new Money(Number::fromFloat(3.5), new Currency('USD');`

The point of this PR was to provide a constructor that can take a string, int or float, so I don't have to do any checking up front.

I'm not particularly attached to the idea though, so no problem if you guys don't like it!

**Comment by sagikazarmark** [comment]
Hm, I see. I'll leave this decision to @frederikbosch

**Comment by frederikbosch** [comment]
@bendavies I get your point, but I stick to my main point: I do not want to have a constructor that becomes a bag of mixed types.

What could be the solution that fits both viewpoints.

```php
Money::fromFloat(3.5);

public static function fromFloat($amount)
{
  if (!is_float($amount)) {
    throw new \InvalidArgumentException();
  }

  return new self((string)Number::fromFloat($amount));
}
```

**Comment by frederikbosch** [comment]
This issue has too many problems. If there is a float, say `3.5`, should that become 3 or 4? Or should we use a rounding mode? But since the method will be static, we have to change more. And since I am totally against making the constructor a bag mixed types, I am going to close it. This is possible, but requires a more comprehensive solution than the current proposition.

**Pull request closed without merging**